src/Development/IDE/Plugin/CodeAction/Rules.hs

```diff
@@ -0,0 +1,59 @@
+module Development.IDE.Plugin.CodeAction.Rules
+  ( rulePackageExports,
+    getVisiblePackageExports
+  )
+where
+
+import Data.HashMap.Strict (fromListWith)
+import qualified Data.HashMap.Strict as Map
+import Data.Text ( Text, pack,)
+import Data.Traversable (forM)
+import Development.IDE.Core.RuleTypes
+import Development.IDE.Core.Shake
+import Development.IDE.GHC.Util
+import Development.IDE.Plugin.CodeAction.RuleTypes
+import Development.IDE.Types.Location
+import Development.Shake
+import HscTypes (hsc_dflags, IfaceExport, mi_exports)
+import LoadIface
+import Maybes
+import Module (Module (..), ModuleName, moduleNameString)
+import Packages (explicitPackages, exposedModules, packageConfigId)
+import TcRnMonad (WhereFrom (ImportByUser), initIfaceLoad)
+import GHC (DynFlags(pkgState))
+import Control.Monad.Trans (MonadTrans(lift))
+
+getVisiblePackageExports :: NormalizedFilePath -> Action [PackageExportsMap]
+getVisiblePackageExports f = fmap (fromMaybe []) $ runMaybeT $ do
+        session <- MaybeT $ use GhcSession f
+        let exposedPkgs = explicitPackages $ pkgState (hsc_dflags $ hscEnv session)
+        traverse (\p -> lift $ use_ (fromUnitId p) f) exposedPkgs
+
+rulePackageExports :: Rules ()
+rulePackageExports = define $ \p file -> do
+  pkgState <- hscEnv <$> use_ GhcSession file
+  let pkg = lookupPackageConfig (packageExportsUnitId p) pkgState
+  case pkg of
+    Nothing -> return ([], Just mempty)
+    Just pkg -> do
+      results <- forM (exposedModules pkg) $ \(mn, _) -> do
+            modIface <-
+              liftIO $ initIfaceLoad pkgState $
```

Comment:
How does GHC normally load interfaces of dependencies? I thought it was somewhat lazy and didn’t need to suck in all interfaces of all dependencies whereas this will force all of them to be loaded.

Reply:
That's exactly right. GHC loads dependencies lazily, which is an obstacle to guessing imports of dependencies that have not been loaded in the session so far. To work around that, we load the dependencies upfront. Note that we only load the *direct* dependencies, and not the whole package db.

Since the package modules are loaded via interface files this is relatively fast and lightweight. If it ends up bloating the GHC session too much, we could look into loading them into a separate session, then throwing it away after harversting all the info.